reframe/utility/sanity.py

```diff
@@ -471,26 +471,32 @@ def assert_reference(val, ref, lower_thres=None, upper_thres=None, msg=None):
     :arg val: The value to check.
     :arg ref: The reference value.
     :arg lower_thres: The lower threshold value expressed as a negative decimal
-        fraction of the reference value.  Must be in [-1, 0]. If ``None``, no
-        lower thresholds is applied.
+        fraction of the reference value.  Must be in [-1, 0] for ref >= 0.0 and
+        in [-inf, 0] for ref < 0.0.
+        If ``None``, no lower thresholds is applied.
     :arg upper_thres: The upper threshold value expressed as a decimal fraction
-        of the reference value.  Must be in [0, 1]. If ``None``, no upper
-        thresholds is applied.
+        of the reference value. Must be in [0, inf] for ref >= 0.0 and
+        in [0, 1] for ref < 0.0.
+        If ``None``, no upper thresholds is applied.
     :returns: ``True`` on success.
     :raises reframe.core.exceptions.SanityError: if assertion fails or if the
         lower and upper thresholds do not have appropriate values.
     """
     if lower_thres is not None:
+        lower_thres_limit = -1 if ref >= 0 else None
         try:
-            evaluate(assert_bounded(lower_thres, -1, 0))
+            evaluate(assert_bounded(lower_thres, lower_thres_limit, 0))
         except SanityError:
-            raise SanityError('invalid low threshold value: %s' % lower_thres)
+            raise SanityError('invalid low threshold value: %s'
+                              % lower_thres) from None
 
     if upper_thres is not None:
+        upper_thres_limit = None if ref >= 0 else 1
         try:
-            evaluate(assert_bounded(upper_thres, 0, 1))
+            evaluate(assert_bounded(upper_thres, 0, upper_thres_limit))
         except SanityError:
-            raise SanityError('invalid high threshold value: %s' % upper_thres)
+            raise SanityError('invalid high threshold value: %s'
+                              % upper_thres) from None
```

Comment:
We split the lines after the binary operator. [PEP8](https://www.python.org/dev/peps/pep-0008/#should-a-line-break-before-or-after-a-binary-operator) prefers splitting before but does not mandate it.